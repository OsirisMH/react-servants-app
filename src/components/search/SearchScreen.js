import React, { useMemo } from 'react';
import queryString from 'query-string';
import { ServantCard } from '../servants/ServantCard';
import { useForm } from '../../hooks/useForm';
import { useLocation } from 'react-router';
import { getServantByName } from '../../selectors/getServantByName';

export const SearchScreen = ({ history }) => {

    const location = useLocation();
    
    const { q = '' } = queryString.parse(location.search);    
    const [ formValues, handleInputChange ] = useForm( { searchText: q } );
    const { searchText } = formValues;
    
    const servantsFiltered = useMemo(() => getServantByName( q ), [q])

    const handleSearch = (e) => {
        e.preventDefault();
        history.push(`?q=${searchText}`)
    };

    return (
        <div>
            <h1>SearchScreen</h1>
            <hr />

            <div className="row">
                <div className="col-5">
                    <h4>Search Form</h4>
                    <hr />

                    <form onSubmit={ handleSearch }>
                        <input 
                            type="text"
                            placeholder="find a servant"
                            className="form-control"
                            name="searchText"
                            autoComplete="off"
                            value={ searchText }
                            onChange={ handleInputChange }
                        />
                        <div className="d-grid gap-2">
                            <button
                                type="submit"
                                className="btn my-1 btn-outline-primary"
                            >
                                Search...
                            </button>
                        </div>
                    </form>
                </div>
                <div className="col-7">
                    <h4>Results</h4>
                    <hr />

                    { 
                        ( q === '' ) ?
                            <div className="alert alert-info">
                                Search a servant
                            </div>
                        : ( servantsFiltered.length === 0 ) &&
                            <div className="alert alert-danger">
                                Servant not found
                            </div>
                    }

                    {/* { ( q === '' )
                        &&
                        <div className="alert alert-info">
                            Search a servant
                        </div>
                    }

                    { ( q !== '' && servantsFiltered.length === 0 )
                        &&
                        <div className="alert alert-danger">
                            Servant not founded
                        </div>
                    } */}

                    {
                        servantsFiltered.map( servant => (
                            <ServantCard
                                key={servant.id}
                                { ...servant }
                            />
                        ))
                    }
                </div>
            </div>
        </div>
    )
};
