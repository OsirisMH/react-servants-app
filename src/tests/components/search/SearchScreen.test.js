import { mount } from "enzyme";
import { MemoryRouter, Route } from "react-router";
import { SearchScreen } from "../../../components/search/SearchScreen";

describe('Pruebas en <SearchScreen />', () => {
   
    test('Debe de mostrarse correctamente con valores por defecto', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']} >
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.alert-info').text().trim() ).toBe('Search a servant');

    });
    
    test('Debe de mostrar Okita y el input con el valor del queryString', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=okita']} >
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        );

        expect( wrapper.find('input').prop('value') ).toBe('okita');
        expect( wrapper ).toMatchSnapshot();

    });
    
    test('Debe de mostrar un error si no se encuentra el servant', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=altria']} >
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        );
        
        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.alert-danger').exists() ).toBe( true );
        expect( wrapper.find('.alert').text().trim() ).toBe('Servant not found');

    });

    test('Debe de llamar el push del history', () => {
        
        const history = {
            push: jest.fn()
        };

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=altria']} >
                <Route
                    path="/search"
                    component={ () => <SearchScreen history={ history } /> }
                />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText',
                value: 'crane'
            }
        });

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });
        
        expect( history.push ).toHaveBeenCalledWith(`?q=crane`);

    });
    
});
