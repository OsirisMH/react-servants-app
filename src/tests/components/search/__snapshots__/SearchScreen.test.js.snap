// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> Debe de mostrar Okita y el input con el valor del queryString 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="find a servant"
          type="text"
          value="okita"
        />
        <div
          className="d-grid gap-2"
        >
          <button
            className="btn my-1 btn-outline-primary"
            type="submit"
          >
            Search...
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="col"
      >
        <div
          className="card mb-3"
          style={
            Object {
              "maxWidth": 540,
            }
          }
        >
          <div
            className="row g-0"
          >
            <div
              className="col-md-4"
            >
              <img
                alt="Okita Sōji"
                className="card-img"
                src="./assets/servants/saber-okita.png"
              />
            </div>
            <div
              className="col-md-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Okita Sōji
                </h5>
                <p
                  className="card-text"
                >
                  Mumyou Sandan-Zuki
                   
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Historical Fact
                  </small>
                </p>
                <a
                  href="/servant/saber-okita"
                  onClick={[Function]}
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen /> Debe de mostrar un error si no se encuentra el servant 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="find a servant"
          type="text"
          value="altria"
        />
        <div
          className="d-grid gap-2"
        >
          <button
            className="btn my-1 btn-outline-primary"
            type="submit"
          >
            Search...
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        Servant not found
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen /> Debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="find a servant"
          type="text"
          value=""
        />
        <div
          className="d-grid gap-2"
        >
          <button
            className="btn my-1 btn-outline-primary"
            type="submit"
          >
            Search...
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a servant
      </div>
    </div>
  </div>
</div>
`;
